File: 24秋招/联想笔试/2.js
class TreeNode {
  constructor(val) {
    this.val = val
    this.children = []
  }
}

function build(n, edges) {
  const nodeList = Array.from({ length: n }, (_, i) => new TreeNode(i + 1))
  for (let i = 1; i < n; i++) {
    nodeList[edges[i - 1] - 1].children.push(nodeList[i])
  }
  return nodeList[0]
}

function dfs(node, depthList) {
  if (!node.children.length) {
    depthList[node.val] = 1
    return 1
  }

  let depth = 0
  for (const child of node.children) {
    depth += dfs(child, depthList)
  }

  depthList[node.val] = depth
  return depth
}

// Test Input
// const input = `8 4
// 1 1 2 2 3 3 3
// 4 2 1 5
// 5 3 2 8`
// var input = ''
// while ((line = readline())) {
//   input += line
// }
// const lines = input.split('\n')
let line = readline()
const [n, Q] = line.split(' ').map(Number)
line = readline()
const edges = line.split(' ').map(Number)
line = readline()
const queries1 = line.split(' ').map(Number)
line = readline()
const queries2 = line.split(' ').map(Number)

const root = build(n, edges)
const depthList = {}
dfs(root, depthList)

let res = 0
for (let i = 0; i < Q; i++) {
  res ^= depthList[queries1[i]] * depthList[queries2[i]]
}

console.log(res)

File: 24秋招/联想笔试/2.py
class TreeNode:
    def __init__(self, val):
        self.val = val
        self.children = []

def foo(n: int, edgeList: list) -> TreeNode:

    nodeList = [TreeNode(i) for i in range(1, n+1)]
    for i, f in enumerate(edgeList, start=1):
        nodeList[f-1].children.append(nodeList[i])
    return nodeList[0]

def dfs(node: TreeNode, depthList: dict) -> int:

    if not node.children: 
        depthList[node.val] = 1
        return 1

    for child in node.children: 
        if node.val not in depthList:
            depthList[node.val] = 0
        depthList[node.val] += dfs(child, depthList)
        
    return depthList[node.val]

if __name__ == "__main__":

    n, Q = map(int, input().split())
    edgeList = list(map(int, input().split()))
    queries1 = list(map(int, input().split()))
    queries2 = list(map(int, input().split()))

    root = foo(n, edgeList)
    depthList = dict()
    dfs(root, depthList)

    res = 0
    for a, b in zip(queries1, queries2):
        res ^= depthList[a] * depthList[b]

    print(res)
